lib/tsconfig/TsConfigLua.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+/* 
+ * File:   TSConfigLua.h
+ * Author: persia
+ *
+ * Created on September 21, 2017, 4:04 PM
+ */
+
+
+#ifndef TSCONFIGLUA_H
+#define TSCONFIGLUA_H
+
+
+#include "Errata.h"
+#include "lua.h"
+#include <unordered_map>
+
+class TsConfigDescriptor {
+   enum class Type { ARRAY, OBJECT, INT, FLOAT, STRING, ENUM };
+   std::string name;
+   std::string description;
+   Type type;
+   std::string type_name;
+};
+
+class TsConfigBase {
+public:
+   TsConfigBase(TsConfigDescriptor const& d) : descriptor(d) {}
+   TsConfigDescriptor const& descriptor;
+
+   virtual ts::Errata loader(lua_State* s) = 0;
+};
+
+template < typename T >
+class TsConfigInt : public TsConfigBase {
+   TsConfigInt(TsConfigDescriptor const& d, int& i) : TsConfigBase(d), ref(i) {}
+   int & ref;
+   ts::Errata loader(lua_State* s) override;
+};
+
+template < typename T >
+class TsConfigString : public TsConfigBase {
+public:
+   TsConfigString(TsConfigDescriptor const& d, std::string& str) : TsConfigBase(d), ref(str) {}
+    std::string& ref;
+    TsConfigString& operator= (const TsConfigString& other)
+    {
+        ref = other.ref;
+        return *this;
+    }
+   ts::Errata loader(lua_State* s) override;
+};
+
+template < typename T, typename E >
+class TsConfigEnum : public TsConfigBase {
+public:
+   TsConfigEnum(TsConfigDescriptor const& d, E& i) : TsConfigBase(d), ref(i) {}
+   E& ref;
+   TsConfigEnum& operator= (const TsConfigEnum& other)
+    {
+        ref = other.ref;
+        return *this;
+    }
+   ts::Errata loader(lua_State* s) override;
+};
+
+class TsConfigArrayDescriptor : public TsConfigDescriptor {
+public:
+   TsConfigArrayDescriptor(TsConfigDescriptor const& d) : item(d) {}
+   TsConfigDescriptor const& item;
+};
+
+class TsConfigEnumDescriptor : public TsConfigDescriptor {
```

Comment:
This should be moved earlier in the file and `TsConfigEnum` should take a this, `TsConfigEnumDescriptor`, as its constructor argument. Unfortunately we'll probably want a method to do the casting from `TsConfigDescriptor&` to `TsConfigEnumDescriptor&`, although if it's generated maybe that's not a concern.